### PR TITLE
Add rustix backend

### PR DESCRIPTION
**Pull request by notgull**
This commit adds a backend based on the rustix crate. The main advantage
of rustix, in addition to greater safety, is that it uses raw syscalls
instead of going through libc.

I've tried to modify the existing libc code as little as possible, in
order to make this change as auditable as possible. I haven't touched
the pthreads code yet, as that's a little tricky.

This exists for discussion purposes.

cc #401

### CODE REVIEW

File: src/use_file.rs [hunk 1]
```diff
@@ -93,6 +115,25 @@ fn wait_until_rng_ready() -> Result<(), Error> {
...
+    loop {
+        match event::poll(&mut pfd, -1) {
+            Ok(_) => return Ok(()),
+            Err(rustix::io::Errno::INTR) => continue,
+            Err(err) => return Err(crate::util_rustix::cvt(err)),
+        }
+    }
+}
+
 struct Mutex(UnsafeCell<libc::pthread_mutex_t>);
```
Comment (briansmith) [inline src/use_file.rs:137]: The most important improvement rustix could make to this code would be replacing the pthreads_mutex_t here with whatever is used by libstd on the target (e.g. futex on Linux/Android).

Comment (notgull) [reply src/use_file.rs]: Good point. I've replaced `Mutex` with one based on futexes, only on Rustix.

File: Cargo.toml [hunk 2]
```diff
@@ -18,14 +18,14 @@ cfg-if = "1"
...
+[target.'cfg(all(unix, not(target_os = "linux")))'.dependencies]
 libc = { version = "0.2.154", default-features = false }
 
 [target.'cfg(target_os = "wasi")'.dependencies]
 wasi = { version = "0.11", default-features = false }
 
-[target.'cfg(any(target_os = "linux", target_os = "android"))'.dependencies]
-rustix = { version = "0.38.0", features = ["event", "fs", "rand", "thread"], default-features = false, optional = true }
+[target.'cfg(target_os = "linux")'.dependencies]
+rustix = { version = "0.38.0", features = ["event", "fs", "rand", "thread"], default-features = false }
```
Comment (briansmith) [inline Cargo.toml:28]: We should only need `event`, `fs`, and `thread` if we're going to use the file-based fallback, but unfortunately that's very hard to specify in Cargo.toml, as you can see by the actual cfg logic in lib.rs.

I guess there must be a reason why Rustix makes those things optional features. (Why?) Would people who don't want those features activated care that `getrandom` would be activating them?

Comment (notgull) [reply Cargo.toml:28]: I don't think so, the features are mostly to reduce code size.

**Comment by newpavlov** [comment]
This PR is outdated, so I will close it. We probably will revisit the `rustix` question while adding support for `none` Linux targets.

**Pull request closed without merging**